esrally/client/synchronous.py

```diff
@@ -211,6 +202,15 @@ def perform_request(
 
             raise HTTP_EXCEPTIONS.get(meta.status, ApiError)(message=message, meta=meta, body=resp_body)
 
+        if self._verified_elasticsearch is None:
+            info = self.transport.perform_request(method="GET", target="/", headers=request_headers)
+            info_meta = info.meta
+            info_body = info.body
+            self._verified_elasticsearch = _ProductChecker.check_product(info_meta.headers, info_body)
+
+            if self._verified_elasticsearch is not True:
+                _ProductChecker.raise_error(self._verified_elasticsearch, info_meta, info_body)
+
```

Comment:
Good catch. I originally moved this when I was testing exceptions because I was getting `UnsupportedProductErrors` even when I expected `ConnectionErrors` etc.

I've refactored it back to before the first request, and to at least check for `ApiErrors` in https://github.com/elastic/rally/pull/1669/commits/f0a56933c0e01db522b9626cfb976a29b962434a.